File: 이진영/프로그래머스/주차 요금 계산.js
function solution(fees, records) {
    const close_time = 23 * 60 + 59;
    let totalTime = new Map();
    let park = new Map();
    let cars = new Set();
    
    let q = [...records];
    while (q.length > 0) {
        const [time, car, type] = q.shift().split(" ");
        const t = getTime(time);
        
        if (type === "IN") {
            park.set(car, t);
            cars.add(car);
        } else {
            totalTime.set(car, (totalTime.get(car) || 0) + t - park.get(car));
            park.delete(car);
        }
    }
    
    for (let [car, in_time] of park.entries()) {
        totalTime.set(car, (totalTime.get(car) || 0) + close_time - in_time);
    }

    return [...cars].sort((a, b) => a - b).map(v => getFee(fees, totalTime.get(v)));
}

const getTime = (time_str) => {
    const [h, m] = time_str.split(":").map(Number);
    return m + 60 * h;
}

const getFee = (fees, time) => {
    if (time <= fees[0]) return fees[1];
    
    return Math.ceil((time - fees[0])/fees[2]) * fees[3] + fees[1];
}

File: 정민선/프로그래머스/주차요금계산.py
from datetime import datetime, timedelta
import math

def solution(fees, records):
    answer = []
    # fees: 기본시간, 기본 요금, 기준 시간, 기준 요금
    basetime, basefee, unittime, unitfee = fees
    fmt = "%H:%M"
    park = dict()
    park_fee = []
    
    for r in records:
        time, car_number, state = map(str, r.split())
        
        if car_number not in park:
            park[car_number] = {'time' : time, 'total': 0}
        
        if state == "IN":
            park[car_number]['time'] = time
        else:
            t1 = datetime.strptime(time, fmt)
            t2 = datetime.strptime(park[car_number]['time'], fmt)

            park[car_number]['total'] += (t1 - t2).seconds//60
            park[car_number]['time'] = None

    for car, info in park.items():
        if info['time']:
            t1 = datetime.strptime("23:59", fmt)
            t2 = datetime.strptime(info['time'], fmt)

            info['total'] += (t1 - t2).seconds//60
            info['time'] = None
    
    for car, info in park.items():
        total = info['total']
        fee = basefee
        
        if total >= basetime:
            fee += math.ceil((total-basetime)/unittime)*unitfee
            
        park_fee.append((car, fee))
        
    park_fee.sort(key=lambda x: (x[0], x[1]))
    
    for car, fee in park_fee:
        answer.append(fee)
        
    return answer